File: Python/calc_redo Covid.py
import pandas as pd
import numpy as np
import pickle
import os
import re

# Read in latest data
covid = pd.read_excel('Input/covid paper info for Elisa_clean.xlsx', sheet_name='4th dataset')

# Pivot wider for co-authors
wider = pd.concat([covid.drop('Co-authors',axis=1),covid['Co-authors'].str.split(", ", expand=True).add_prefix('author')],axis=1)

# Find how many co-authors +1 for main author
string = ''.join(wider.columns)
pos = int(max(re.findall(r'author([0-9]+)', string))) +1

wider['author'+ str(pos)]=wider['Forename']+' '+wider['Surname']
wider.drop(['Forename','Surname'], axis=1, inplace=True)
author_cols = [cols for cols in wider.columns if cols.startswith('author')]
id_vars = [cols for cols in wider.columns if not cols.startswith('author')]

longer = wider.melt(id_vars=id_vars, value_vars=author_cols, value_name='author').dropna(subset=['author']).drop('variable',axis=1)

# Gender, Institution, paper titles are for lead authors
lead_author = covid.iloc[:,0:3]
lead_author['lead_author'] = lead_author['Forename'] + ' ' + lead_author['Surname']

lead_author.drop(columns = ['Forename', 'Surname']).rename(columns={"Unnamed: 0":"gender"})

new = pd.concat([longer.drop('author',axis=1), 
                longer['author'].str.split(' and ', expand=True).add_prefix('author')], axis=1)

# +
author_cols = [cols for cols in new.columns if cols.startswith('author')]

id_vars = [cols for cols in new.columns if not cols.startswith('author')]


# -

new = new.melt(id_vars=id_vars, value_vars=author_cols, value_name = 'author').dropna(subset=['author']).drop(
    'variable',axis=1)

new['author'] = new['author'].str.strip()

new.shape

new.drop_duplicates(inplace=True)

new.shape

new.columns = ['gender','affiliation', 'title', 'date', 'decision', 'author']

new.columns

new.reset_index(drop=True, inplace=True)

new['last_name'] = new['author'].str.split(' ').str[-1]
new['first_name'] = new['author'].str.split(' ').str[0]

new

new['gender'] = np.where(new['gender']=='M', 'male', 
                                      np.where(new['gender']=='F', 'female', None))

new

covid['lead_author']=1

covid['author'] = covid['Forename']+ ' ' +covid['Surname']

indicator = covid[['author','Paper Title','lead_author']]

indicator.columns = ['author','title','lead_author']

new = pd.merge(left=new, right=indicator, how='left')

new['affiliation'] = np.where(new['lead_author'].isna(), None, new['affiliation'])

new

new.to_pickle('covid/covid_clean_winter.pickle')

File: R/calc-redo covid.R
library(here)
library(tidyverse)
library(stringi)
library(lubridate)

file = 'covid paper info for Elisa_clean.xlsx'
sheets = c('1st data set', '2nd data set', '3rd dataset', '4th dataset')
output_file = paste0('covid', 1:4, '.RDS')

clean_covid_sheet <- function(filename, sheetname, outputname){

    # Read in dataset
    df <- readxl::read_xlsx(here('Input', 'COVID', filename), sheet = sheetname)
    
    colnames(df) <- c('gender', 'first_name', 'last_name', 'affiliation',
                      'title', 'coauthors', 'date', 'decision')
    
    max_authors = max(df %>% 
                        # Check number of authors
                        mutate(n_authors = str_count(`coauthors`, pattern = ",")) %>% 
                        pull(n_authors), na.rm=TRUE)+1
    
    df_clean <- df %>% 
      # replace and
      mutate(coauthors = str_replace_all(coauthors, ' and', ',')) %>% 
      mutate(coauthors = str_replace_all(coauthors, ';', ',')) %>% 
      # split authors by ,
      separate(coauthors, paste0('authors',1:max_authors), sep=',') %>% 
      # set lead author as author100
      mutate(author100 = paste(first_name, last_name)) %>% 
      # pivot longer
      pivot_longer(cols=starts_with('author'), values_to = 'author') %>% 
      filter(!is.na(author)) %>% 
      # set lead author descriptions
      select(-first_name, -last_name) %>% 
      mutate(gender = ifelse(name=='author100', gender, NA),
             affiliation = ifelse(name=='author100', affiliation, NA)) %>% 
      # copy out extra affiliations in brackets
      mutate(affiliation = ifelse(grepl('\\(', author), str_extract(author, '\\([a-zA-Z\\- ]+\\)'),
                                  affiliation),
             affiliation = gsub('[()]', '', affiliation)) %>% 
      # clean up names
      mutate(author = str_replace(author, "\\([a-zA-Z\\- ]+\\)", ''),
             author = str_trim(author), 
             author = str_to_title(author), 
             author = stri_trans_general(author, "Latin-ASCII")) %>% 
      # drop any blanks
      filter(author != '') %>% 
      select(-name) %>% 
      # create first and last name
      mutate(first_name = word(author, start = 1 , end = 1),
             last_name = word(author, start=-1, end=-1)) %>% 
      # format date
      mutate(date = ymd(date)) %>% 
      # format gender
      mutate(gender = ifelse(gender == 'M', 'Male',
                             ifelse(gender == 'F', 'Female', NA)))
    
    write_rds(df_clean, here('Output', outputname))

}

read_covid <- function(myfile, mysheets, my_output_file){

    for(i in 1:length(sheets)){
      clean_covid_sheet(myfile, mysheets[i], my_output_file[i])
    }
    
    covid_files = list.files(here('Output'), pattern = 'covid*', full.names=TRUE)
    df_list = lapply(covid_files, read_rds)
    df_merge = reduce(df_list, bind_rows)
    
    # Clean out duplicates
    df_clean = df_merge %>% 
      group_by(title, first_name, last_name, date) %>% 
      mutate(gender = gender[which(!is.na(gender))[1]],
             affiliation = affiliation[which(!is.na(affiliation))[1]]) %>% 
      ungroup() %>% 
      distinct()
    
    return(df_clean)
}



